### PR TITLE
Enhance Extract Genomic DNA

### PR DESCRIPTION
**Pull request by gregvonkuster**
- allow selection of a field delimiter for the Fasta description when output is selected to be Fasta

### CODE REVIEW

**Comment by gregvonkuster** [comment]
I will have an addition to this coming shortly.

**Comment by gregvonkuster** [comment]
This is now ready for review, thanks!

File: tools/extract_genomic_dna/extract_genomic_dna.xml [hunk 1]
```diff
@@ -78,6 +105,8 @@
             <param name="interpret_features" value="yes"/>
             <param name="index_source" value="cached"/>
             <param name="out_format" value="fasta"/>
```
Comment (nsoranzo) [inline tools/extract_genomic_dna/extract_genomic_dna.xml]: `s/out_format/output_format/` on all tests

File: tools/extract_genomic_dna/extract_genomic_dna.xml [hunk 2]
```diff
@@ -60,15 +66,36 @@
...
+                        <option value="char_delimited">character delimited field values</option>
+                    </param>
+                    <when value="bedtools_getfasta_default"/>
+                    <when value="char_delimited">
+                        <param name="fasta_header_delimiter" type="select" label="Select fasta header field delimiter">
+                            <option value="underscore" selected="True">underscore (_)</option>
+                            <option value="semicolon">semicolon (;)</option>
+                            <option value="comma">comma (,)</option>
+                            <option value="tilda">tilda (~)</option>
+                            <option value="vetical_bar">vertical bar (|)</option>
```
Comment (nsoranzo) [inline tools/extract_genomic_dna/extract_genomic_dna.xml]: s/vetical_bar/vertical_bar/

File: tools/extract_genomic_dna/extract_genomic_dna_utils.py [hunk 3]
```diff
@@ -292,6 +292,39 @@ def convert_to_twobit(reference_genome):
...
+    # Return a specified fasta header delimiter.
+    if delimiter == 'underscore':
+        return '_'
+    if delimiter == 'semicolon':
+        return ';'
+    if delimiter == 'comma':
+        return ','
+    if delimiter == 'tilda':
+        return '~'
+    if delimiter == 'vetical_bar':
```
Comment (nsoranzo) [inline tools/extract_genomic_dna/extract_genomic_dna_utils.py]: s/vetical_bar/vertical_bar/

File: tools/extract_genomic_dna/extract_genomic_dna_utils.py [hunk 4]
```diff
@@ -292,6 +292,39 @@ def convert_to_twobit(reference_genome):
...
+
+def get_fasta_header_delimiter(delimiter):
+    # Return a specified fasta header delimiter.
+    if delimiter == 'underscore':
+        return '_'
+    if delimiter == 'semicolon':
+        return ';'
+    if delimiter == 'comma':
+        return ','
+    if delimiter == 'tilda':
```
Comment (nsoranzo) [inline tools/extract_genomic_dna/extract_genomic_dna_utils.py]: s/tilda/tilde/

File: tools/extract_genomic_dna/extract_genomic_dna.xml [hunk 5]
```diff
@@ -60,15 +66,36 @@
...
+                        <option value="bedtools_getfasta_default" selected="True">bedtools getfasta default</option>
+                        <option value="char_delimited">character delimited field values</option>
+                    </param>
+                    <when value="bedtools_getfasta_default"/>
+                    <when value="char_delimited">
+                        <param name="fasta_header_delimiter" type="select" label="Select fasta header field delimiter">
+                            <option value="underscore" selected="True">underscore (_)</option>
+                            <option value="semicolon">semicolon (;)</option>
+                            <option value="comma">comma (,)</option>
+                            <option value="tilda">tilda (~)</option>
```
Comment (nsoranzo) [inline tools/extract_genomic_dna/extract_genomic_dna.xml]: s/tilda/tilde/ (twice)

**Comment by gregvonkuster** [comment]
Thanks for the review @nsoranzo  - fixes are included here: https://github.com/gregvonkuster/tools-iuc/commit/ea98f16396fdf9d7e16df3e1d0c818574f5b0c39

The main reason for these issues is that I've not been able to test tools with planemo for quite some time now.  Do you know if planemo is working for testing tools?  If it is, would you be able to help me out - perhaps I'm not doing things correctly.

I'm running planemo from a virtual env outside of the Galaxy environment, and I'm running this version:

(venv)dot1x-cb-16:tool gvk$ planemo --version
planemo, version 0.23.0

I've tried testing tools with this command where my ~/.planemo.yml file points to a Galaxy instance running the dev branch and tests do not run - this error is displayed at the end:

planemo test extract_genomic_na.xml

axy_test/structures.py", line 199, in _xunit_root
    return self.xunit_tree.getroot()
AttributeError: 'Element' object has no attribute 'get root'

I've also tried this command that has pleanmo install its won Galaxy instance for testing:

planemo t --install_galaxy extract_genomic_na.xml
which results in the same error:

  File "/Users/gvk/work/virtual_env/venv/lib/python2.7/site-packages/planemo/galaxy_test/structures.py", line 207, in xunit_testcase_elements
    return xunit_t_elements_from_root(self._xunit_root)
  File "/Users/gvk/work/virtual_env/venv/lib/python2.7/site-packages/planemo/galaxy_test/structures.py", line 199, in _xunit_root
    return self.xunit_tree.getroot()
AttributeError: 'Element' object has no attribute 'get root'

The exceptions thrown by testing with planemo evolve often - it seems like no specific error remains for very long.  New ones just pop up regularly.

If planemo is still being used for testing tools with Jenkins and other tools, I must be doing something wrong.  I'm just not quite sure what it is.

Any help would be very much appreciated.  Thanks!!  ;)

**Comment by bgruening** [comment]
@gregvonkuster sorry very less time today, but try run planemo without a virtual-env. Also have a look at the `.travis.yaml` file. We have integrated travis/planemo testing for non-blacklisted tools here in the IUC repo. So yes it is working and your issues are somehow caused by your Python environment. Also make sure you have Galaxy 16.01 as planemo test target.

**Comment by gregvonkuster** [comment]
@bgruening Thanks!  I'll take a look at what you've advised and see if I can get things working.

File: tools/extract_genomic_dna/extract_genomic_dna.xml [hunk 6]
```diff
@@ -77,21 +104,25 @@
             <param name="input" value="1.bed" dbkey="hg17" ftype="bed" />
             <param name="interpret_features" value="yes"/>
             <param name="index_source" value="cached"/>
```
Comment (nsoranzo) [inline tools/extract_genomic_dna/extract_genomic_dna.xml]: `s/index_source/reference_genome_source/` in all tests.

I think all tests using "cached" will fail.

File: tools/extract_genomic_dna/extract_genomic_dna.xml [hunk 7]
```diff
@@ -115,7 +148,9 @@
             <param name="interpret_features" value="no"/>
             <param name="index_source" value="history"/>
             <param name="ref_file" value="tophat_in1.fasta"/>
```
Comment (nsoranzo) [inline tools/extract_genomic_dna/extract_genomic_dna.xml]: `s/ref_file/reference_genome/` here and in the next test.

File: tools/extract_genomic_dna/extract_genomic_dna.xml [hunk 8]
```diff
@@ -77,21 +104,25 @@
             <param name="input" value="1.bed" dbkey="hg17" ftype="bed" />
             <param name="interpret_features" value="yes"/>
             <param name="index_source" value="cached"/>
-            <param name="out_format" value="fasta"/>
+            <param name="output_format" value="fasta"/>
+            <param name="fasta_header_type" value="char_delimited"/>
+            <param name="fasta_header_delimiter" value="underscore"/>
             <output name="out_file1" file="extract_genomic_dna_out1.fasta" compare="contains" />
```
Comment (nsoranzo) [inline tools/extract_genomic_dna/extract_genomic_dna.xml]: `s/out_file1/output/` in all tests.

**Comment by bgruening** [comment]
@gregvonkuster if you merge my small PR travis will test your PR with all dependencies :)

**Comment by gregvonkuster** [comment]
@nsoranzo Should I still change all tests that use cached reference genomes to now use reference genomes from the history?  These tests were all inherited from the original version of the tool that is still in the Galaxy core.  I'm happy to change these if they will not pass tests with the cached genomes.  Just want to make sure I should do this after Bjorn's patch.  Thanks!

**Comment by bgruening** [comment]
I would vote for it also I think there is a secret way to test these cached databases, I think @peterjc was working/using it in the blast-wrappers. I really want to have tools, especially new ones under testing, finally.

**Comment by peterjc** [comment]
The Tool Shed framework didn't support it, but the Galaxy tests (and via planemo) can use test ``test-data/*.loc`` files, and we are using that for the BLAST+ wrappers - see https://github.com/peterjc/galaxy_blast/issues/53

**Comment by gregvonkuster** [comment]
Can anyone point me to small reference fasta datasets that I can use for these tests?  If all current tests are needed, I need small fasta reference files for hg17 and droPer1 at least.  Thanks.

**Comment by bgruening** [comment]
@gregvonkuster it's too large.
http://hgdownload.cse.ucsc.edu/goldenPath/droPer1/bigZips/ over 50MB. You could invent some `foo` genome db-key and include your own small fasta file maybe? get rid of hg17 and droPer1 if you want to test this.

**Comment by gregvonkuster** [comment]
@bgruening @nsoranzo Functional tests are now all using reference genomes from the history in this change: https://github.com/gregvonkuster/tools-iuc/commit/b74c3cb956577cdaa3d660a8661e48c1ddbb7b16
Thanks for all the help on this!  ;)

**Comment by bgruening** [comment]
@gregvonkuster thanks! 3/4 are working!

```
Traceback (most recent call last):

  File "/tmp/tmpUVe7la/galaxy-dev/test/functional/test_toolbox.py", line 289, in test_tool

    self.do_it( td )

  File "/tmp/tmpUVe7la/galaxy-dev/test/functional/test_toolbox.py", line 73, in do_it

    raise e

JobOutputsError: History item  different than expected, difference (using contains):

( /home/travis/build/galaxyproject/tools-iuc/tools/extract_genomic_dna/test-data/extract_genomic_dna_out3.gff v. /tmp/tmpUVe7la/tmp/tmpLJ2ySiextract_genomic_dna_out3.gff )

Failed to find 'chr1	147962192	147962580	CCDS989.1_cds_0_0_chr1_147962193_r	0	-	ACTTGATCCTGCTCCCTCGGTGTCTGCATTGACTCCTCATGCTGGGACTGGACCCGTCAACCCCCCTGCTCGCTGCTCACGTACCTTCATCACTTTTAGTGATGATGCAACTTTCGAGGAATGGTTCCCCCAAGGGCGGCCCCCAAAAGTCCCTGTTCGTGAGGTCTGTCCAGTGACCCATCGTCCAGCCCTATACCGGGACCCTGTTACAGACATACCCTATGCCACTGCTCGAGCCTTCAAGATCATTCGTGAGGCTTACAAGAAGTACATTACTGCCCATGGACTGCCGCCCACTGCCTCAGCCCTGGGCCCCGGCCCGCCACCTCCTGAGCCCCTCCCTGGCTCTGGGCCCCGAGCCTTGCGCCAGAAAATTGTCATTAAATGA' in history data. (lines_diff=0):
```

**Comment by gregvonkuster** [comment]
Sorry!  @bgruening hopefully this fixes the test.  https://github.com/gregvonkuster/tools-iuc/commit/aa43aa341808b21283b34c9bb0e825d20103f963

**Comment by bgruening** [comment]
Fantastic!

**Comment by gregvonkuster** [comment]
Thanks so much @bgruening !!

**Comment by gregvonkuster** [comment]
When convenient, could this new version 3.0.2 of this Extract Genomic DNA tool be uploaded to the MTS?  Thanks!!  ;)

**Comment by bgruening** [comment]
Done!

**Comment by gregvonkuster** [comment]
Thanks @bgruening

**Pull request merged**